pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,520 @@
+		if err != nil {
+			klog.ErrorS(err, "Parse certificate from file error", "file", certPath)
+		} else {
+			return certPEMBytes, certificates, nil
+		}
+	}
+	return nil, nil, nil
+}
+
+func (c *Controller) configureOVS() error {
+	// calculate the hash of current configurations and save it to OVS DB to trigger updates.
+	hasher := sha256.New()
+	existing := c.certificateKeyPair.Load().(*certificateKeyPair)
+	hasher.Write(existing.rawCACert)
+	hasher.Write(existing.rawCert)
+	hasher.Write(existing.rawKey)
+	hash := hex.EncodeToString(hasher.Sum(nil))
+
+	ovsConfig := map[string]interface{}{
+		"certificate":      c.certPath,
+		"private_key":      c.privateKeyPath,
+		"ca_cert":          c.caPath,
+		"certificate_hash": hash,
+	}
+	oldConfig, err := c.ovsBridgeClient.GetOVSOtherConfig()
+	if err != nil {
+		return err
+	}
+	// delete the old config if needed
+	toDelete := make(map[string]interface{})
+	toAdd := make(map[string]interface{})
+	for k, v := range ovsConfig {
+		oldValue, ok := oldConfig[k]
+		if ok && oldValue != v {
+			toDelete[k] = oldConfig[k]
+		}
+		if oldValue == v {
+			continue
+		}
+		toAdd[k] = v
+	}
+	if len(toDelete) > 0 {
+		if err := c.ovsBridgeClient.DeleteOVSOtherConfig(toDelete); err != nil {
+			return err
+		}
+	}
+	if len(toAdd) == 0 {
+		return nil
+	}
+	klog.InfoS("Updating OVS configurations for IPsec certificates")
+	return c.ovsBridgeClient.AddOVSOtherConfig(ovsConfig)
+}
+
+func newCSR(csrName, commonName string, privateKey crypto.PrivateKey) (*csrv1.CertificateSigningRequest, error) {
+	subject := &pkix.Name{
+		CommonName:   commonName,
+		Organization: []string{"antrea.io"},
+	}
+	csrBytes, err := certutil.MakeCSR(privateKey, subject, []string{commonName}, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &csrv1.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: csrName,
+		},
+		Spec: csrv1.CertificateSigningRequestSpec{
+			Request:    csrBytes,
+			SignerName: signerName,
+			Usages:     []v1.KeyUsage{v1.UsageIPsecTunnel},
+		},
+	}, nil
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	// load the existing certificates from files before starting the rotation goroutine.
+	if err := c.syncConfigurations(); err != nil {
+		klog.ErrorS(err, "Sync configurations error")
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	go wait.Until(func() {
+		if err := c.watchFileChanges(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to watch cert and key file, will retry later")
+		}
+	}, time.Minute, stopCh)
+
+	go wait.Until(func() {
+		if err := c.rotateCertificates(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to rotate certificates")
+		}
+	}, time.Minute, stopCh)
+	<-stopCh
+}
+
+func (c *Controller) rotateCertificates(endCh <-chan struct{}) error {
+	for {
+		currentCert := c.certificateKeyPair.Load().(*certificateKeyPair)
```

Comment:
Isn't `c.certificateKeyPair` empty when agent runs this mode first time and no existing certificate exist in file?